pkg/controller/labelidentity/label_group_index.go

```diff
@@ -0,0 +1,503 @@
+func selectorItemKeyFunc(obj interface{}) (string, error) {
+	sItem, ok := obj.(*selectorItem)
+	if !ok {
+		return "", fmt.Errorf("object is not of type *selectorItem: %v", obj)
+	}
+	return sItem.getKey(), nil
+}
+
+func newSelectorItemStore() cache.Indexer {
+	indexers := cache.Indexers{
+		cache.NamespaceIndex: func(obj interface{}) ([]string, error) {
+			sItem, ok := obj.(*selectorItem)
+			if !ok {
+				return []string{}, nil
+			}
+			// sItem.Selector.Namespace == "" means it's a cluster scoped selector, we index it as it is.
+			return []string{sItem.selector.Namespace}, nil
+		},
+		policyIndex: func(obj interface{}) ([]string, error) {
+			sItem, ok := obj.(*selectorItem)
+			if !ok {
+				return []string{}, nil
+			}
+			return sItem.policyKeys.List(), nil
+		},
+	}
+	return cache.NewIndexer(selectorItemKeyFunc, indexers)
+}
+
+// LabelIdentityIndex implements Interface.
+type LabelIdentityIndex struct {
+	lock sync.RWMutex
+	// labelIdentities stores all labelIdentityMatches, with the normalized labels of LabelIdentity as map key.
+	labelIdentities map[string]*labelIdentityMatch
+	// labelIdentityNamespaceIndex is an index from Namespace to LabelIdentity keys in that Namespace.
+	labelIdentityNamespaceIndex map[string]sets.String
+	// selectorItems stores all selectorItems, indexed by Namespace and policy keys.
+	selectorItems cache.Indexer
+
+	eventChan chan string
+	// eventHandlers is a list of callbacks registered for policies to be re-processed due to
+	// LabelIdentity events.
+	eventHandlers []eventHandler
+
+	// synced stores a boolean value, which tracks if the LabelIdentityIndex has been initialized with
+	// the full lists of LabelIdentities.
+	synced *atomic.Value
+}
+
+func NewLabelIdentityIndex() *LabelIdentityIndex {
+	synced := &atomic.Value{}
+	synced.Store(false)
+	index := &LabelIdentityIndex{
+		labelIdentities:             map[string]*labelIdentityMatch{},
+		labelIdentityNamespaceIndex: map[string]sets.String{},
+		selectorItems:               newSelectorItemStore(),
+		eventChan:                   make(chan string, eventChanSize),
+		eventHandlers:               []eventHandler{},
+		synced:                      synced,
+	}
+	return index
+}
+
+func (i *LabelIdentityIndex) updateSelectorItem(sItem *selectorItem, updateType selectorItemUpdateEvent, updateKey string) {
+	labelIdentities, policies := sItem.labelIdentityKeys, sItem.policyKeys
+	switch updateType {
+	case selectorMatchedLabelAdd:
+		labelIdentities.Insert(updateKey)
+	case selectorMatchedLabelDelete:
+		labelIdentities.Delete(updateKey)
+	case selectorMatchedPolicyAdd:
+		policies.Insert(updateKey)
+	case selectorMatchedPolicyDelete:
+		policies.Delete(updateKey)
+	}
+	// Construct a new selectorItem since objects got from ThreadSafeStore should be
+	// read-only. Indexers will break otherwise.
+	newSelectorItem := &selectorItem{
+		selector:          sItem.selector,
+		labelIdentityKeys: labelIdentities,
+		policyKeys:        policies,
```

Comment:
The above code updates policies in-place, which affects the stored object. I think the policy index would be broken and getting items by that index would not work as expected.

Reply:
Are you worried about the concurrent reads by policy index? Because otherwise after the function is executed the original selectorItem will be replaced by the new object and the store should re-index just fine?

Reply:
Refactored though to make a copy of sets.String

Reply:
No, I was worried about the correctness of the index. Treating objects returned by index is not only for concurrent reads but also for index maintenance: it relies on the old object to know how to clear previous indices when an object is updated, you could try the following code to see the consequence of updating an indexed value in-place:
```
func TestIndex(t *testing.T) {
	type Item struct {
		key    string
		values sets.String
	}
	index := cache.NewIndexer(func(obj interface{}) (string, error) {
		return obj.(*Item).key, nil
	}, cache.Indexers{"value": func(obj interface{}) ([]string, error) {
		return obj.(*Item).values.UnsortedList(), nil
	}})

	index.Add(&Item{
		key:    "k1",
		values: sets.NewString("v1", "v2"),
	})

	obj, exists, err := index.GetByKey("k1")
	assert.NoError(t, err)
	assert.True(t, exists)
	item := obj.(*Item)
	key := item.key
	values := item.values
	values.Delete("v1")
	updatedItem := &Item{
		key:    key,
		values: values,
	}
	index.Update(updatedItem)

	objs, err := index.ByIndex("value", "v1")
	assert.NoError(t, err)
        assert.Equal(t, 0, len(objs), "The item should not be returned as v1 is no longer one of its values")
}
```

Reply:
Got it. Thanks for the explaination.